lib/Scheduler/sched_region.cpp

```diff
@@ -164,319 +156,302 @@ FUNC_RESULT SchedRegion::FindOptimalSchedule( //TODO: CHIPPIE: Add helper functi
-           bestCost_);
+      Logger::Info(
+          "Bypassing optimal scheduling due to zero time limit with cost %d",
+          bestCost_);
     } else {
-       Logger::Info("The initial schedule of length %d and cost %d is optimal.",
+      Logger::Info("The initial schedule of length %d and cost %d is optimal.",
                    bestSchedLngth_, bestCost_);
     }
 
     if (rgnTimeout != 0) {
-      bool optimalSchedule = initialScheduleOptimal || (rslt == RES_SUCCESS);
-       Logger::Info("Best schedule for DAG %s has cost %d and length %d. The "
+      bool optimalSchedule = isLstOptml || (rslt == RES_SUCCESS);
+      Logger::Info("Best schedule for DAG %s has cost %d and length %d. The "
                    "schedule is %s",
                    dataDepGraph_->GetDagID(), bestCost_, bestSchedLngth_,
                    optimalSchedule ? "optimal" : "not optimal");
     }
-  }
-
-  //TODO: CHIPPIE: Everything hereafter, should it run for _all_ of the schedulers? Regardless of which ones are enabled?
-  //I.e. are the following blocks of code BB specific? (Doesn't look it...)
 
 #ifdef IS_DEBUG_PRINT_PERP_AT_EACH_STEP
-  Logger::Info("Printing PERP at each step in the schedule.");
+    Logger::Info("Printing PERP at each step in the schedule.");
 
-  int costSum = 0;
-  for (int i = 0; i < dataDepGraph_->GetInstCnt(); ++i) {
-    Logger::Info("Cycle: %lu Cost: %lu", i, bestSched_->GetSpillCost(i));
-    costSum += bestSched_->GetSpillCost(i);
-  }
-  Logger::Info("Cost Sum: %lu", costSum);
+    int costSum = 0;
+    for (int i = 0; i < dataDepGraph_->GetInstCnt(); ++i) {
+      Logger::Info("Cycle: %lu Cost: %lu", i, bestSched_->GetSpillCost(i));
+      costSum += bestSched_->GetSpillCost(i);
+    }
+    Logger::Info("Cost Sum: %lu", costSum);
 #endif
 
-  if (SchedulerOptions::getInstance().GetString(
-          "SIMULATE_REGISTER_ALLOCATION") != "NO") {
-    //#ifdef IS_DEBUG
-    RegAlloc_(bestSched, lstSched);
-    //#endif
+    if (SchedulerOptions::getInstance().GetString(
+            "SIMULATE_REGISTER_ALLOCATION") != "NO") {
+      //#ifdef IS_DEBUG
+      RegAlloc_(bestSched, InitialSchedule);
+      //#endif
+    }
+
+    enumTime = Utilities::GetProcessorTime() - enumStart;
+    stats::enumerationTime.Record(enumTime);
   }
 
-  enumTime = Utilities::GetProcessorTime() - enumStart;
-  stats::enumerationTime.Record(enumTime);
+  // Step 5: Run ACO if schedule from enumerator is not optimal
+  if (bestCost_ != 0 && AcoAfterEnum) {
+    Logger::Info("Final cost is not optimal, running ACO.");
+    InstSchedule *AcoAfterEnumSchedule =
+        new InstSchedule(machMdl_, dataDepGraph_, vrfySched_);
+    if (AcoAfterEnumSchedule == NULL)
+      Logger::Fatal("Out of memory");
+
+    FUNC_RESULT acoRslt = runACO(AcoAfterEnumSchedule, bestSched);
+    if (acoRslt != RES_SUCCESS) {
+      Logger::Info("Running final ACO failed");
+      delete AcoAfterEnumSchedule;
```

Comment:
Won't this cause a segfault on line 437?

Reply:
Ah! I see the issue. Would this fix it?
```c++
    if (acoRslt != RES_SUCCESS) {
      Logger::Info("Running final ACO failed");
      delete AcoAfterEnumSchedule;
    } else {
      InstCount AcoAfterEnumCost = AcoAfterEnumSchedule->GetCost();
      if (AcoAfterEnumCost < bestCost_) {
        InstCount AcoAfterEnumLength = AcoAfterEnumSchedule->GetCrntLngth();
        InstCount imprvmnt = bestCost_ - AcoAfterEnumCost;
        Logger::Info("ACO found better schedule with length=%d, spill cost = %d, "
                     "tot cost = %d, cost imp=%d.",
                     AcoAfterEnumLength, AcoAfterEnumSchedule->GetSpillCost(),
                     AcoAfterEnumCost, imprvmnt);
        bestSched_ = bestSched = AcoAfterEnumSchedule;
        bestCost_ = AcoAfterEnumCost;
        bestSchedLngth_ = AcoAfterEnumLength;
      } else {
        Logger::Info("ACO was unable to find a better schedule.");
        delete AcoAfterEnumSchedule;
      }
    }
```

Reply:
Looks good to me.

Reply:
Updated.